test/gpu_reference_kernel.cpp

```diff
@@ -518,9 +560,15 @@ struct gpu_reference_conv_2d : gpu_reference_kernel_base
                                                nullptr);
                 EXPECT(status == CL_SUCCESS);
 #elif MIOPEN_BACKEND_HIP
+                // TODO(Amber): copy output before computation because output may
```

Comment:
[Q] Can you please explain a bit more why this TODO is necessary? Are you sure that the word "indices" is adequate in this context?

Reply:
So output is non-packed, i.e. larger than what's needed for convolution and may have data that is not touched by convolution operation. In order to verify the output, GPU buffer should have the same data in the `untouched` parts as the CPU, which is achieved by copying the output to the GPU before the convolution is invoked.

Reply:
Have we made sure that untouched areas in the source buffer are zeroed? The random data may contain INF or NAN, etc., and this will spoil the comparison results.

Reply:
@amberhassaan 
> So output is non-packed, i.e. larger than what's needed for convolution and may have data that is not touched by convolution operation. In order to verify the output, GPU buffer should have the same data in the untouched parts as the CPU, which is achieved by copying the output to the GPU before the convolution is invoked.

Please put this explanation into C++ comments.

In fact I would prefer zeroing both reference and output buffers prior computations, and _only if_ the tensors are non-packed. There is some MIOpen kernel that can do this job quickly. Is it possible to do?

Reply:
Tensor-layout-aware may help to compare only changed parts, at least as I understood it. But we also want to check that unchanged parts are actually unchanged, and it means that zero-filled tensors are not so good and fp-comparison is not a good idea too. Ideally it must compare the untouched parts as a binary data and doesn't care about all of those `INF`s and `NAN`s.
N.B. as far as I know current random functions do not generate `INF`s and `NAN`s, but there are some direct `NAN` initialiations.

Reply:
@atamazov , @CAHEK7, good suggestions. I find layout-aware verification a bit much given the deadlines for this work. I see two ways forward (and I prefer 1)
1. Fill the output tensor with random integers (that's how we generate inputs). Our verification (against cpu code) compares all data elements (including those not touched by convolution). 
2. Zero-out the entire tensor, which is not as robust as @CAHEK7 pointed out because we're relying on zero being a special value.

Reply:
@CAHEK7 
> But we also want to check that unchanged parts are actually unchanged

Why? AFAIU the gaps in the real life may contain any garbade.

Reply:
@amberhassaan 
> because we're relying on zero being a special value.

Zero is and will forever be a special value in all IEEE-754 compliant FP numbers and their extensions, as well in all two's complement integers. Are we going to support other number formats in visible future?

Reply:
@amberhassaan If you copy reference output buffer into device output buffer before convolution, then verification will pass even if convolution did nothing.

Reply:
@amberhassaan Aha, now I see. Well, I still think that zeroing is better, but I am Ok with randomizing gaps as well.
[resolved]